### DISCUSSION TITLE
What's the recommended way for installing Chromium on AL2023?

### DISCUSSION THREAD

**Discussion by trivikr** [amazonlinux/amazon-linux-2023#417] [Q&A]
When I search for how to install Chromium on Amazon Linux, the [top result](https://stackoverflow.com/a/72077612) is to use `amazon-linux-extras` as follows:

```console
# Enable and install Extra Packages for Enterprise Linux by running the command
sudo amazon-linux-extras install epel -y

# Post installing all the extra packages successfully, Install chromimum as usual
sudo yum install -y chromium
```

This solution does not work in AL2023, as it does not support amazon-linux-extras.

What's the recommended way for installing Chromium on AL2023?

**Comment by trivikr** [comment]
One solution which worked for me it to install Chromium directly from RPM

```console
$ export BROWSERS_SRC_DIR="/usr/src/browsers" && mkdir -p $BROWSERS_SRC_DIR

$ curl https://dl.google.com/linux/direct/google-chrome-stable_current_x86_64.rpm \
    --output $BROWSERS_SRC_DIR/google-chrome-stable_current_x86_64.rpm
    
$ yum install -y -q $BROWSERS_SRC_DIR/google-chrome-stable_current_x86_64.rpm

$ /usr/bin/google-chrome-stable --version
```

**Reply by idclipr** [reply]
Excellent save! Note that this installs Chrome, not Chromium. This is also a problem with the latest version of Amazon Linux 2 "standard:5.0" used in CodeBuild.

**Reply by rvinas** [reply]
I'm getting this error:
```
Error: 
 Problem: conflicting requests
  - package google-chrome-stable-117.0.5938.88-1.x86_64 does not have a compatible architecture
  - nothing provides ld-linux-x86-64.so.2()(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides ld-linux-x86-64.so.2(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides ld-linux-x86-64.so.2(GLIBC_2.3)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.10)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.11)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.14)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.15)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.16)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.3)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.3.2)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.3.3)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.3.4)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.4)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.6)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.7)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.8)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libc.so.6(GLIBC_2.9)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libdl.so.2(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libm.so.6(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libpthread.so.0(GLIBC_2.12)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libpthread.so.0(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libpthread.so.0(GLIBC_2.3.2)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libpthread.so.0(GLIBC_2.3.3)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides libpthread.so.0(GLIBC_2.3.4)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
  - nothing provides librt.so.1(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
```

**Reply by trivikr** [reply]
@rvinas The solution is verified to be working on AL2023, and we have set it up in our CI.

On 2023-09-18, I created a fresh EC2 instance on AL2023, and ran the commands and confirmed that the version is  `Google Chrome 117.0.5938.88`

Are you using some different Operating System, and not AL2023?

**Reply by ChrisLane** [reply]
Even more direct: `sudo dnf -y https://dl.google.com/linux/direct/google-chrome-stable_current_x86_64.rpm`

**Reply by jkruse14** [reply]
Slight tweak on the answer from @ChrisLane - `install` is needed in the command:

`dnf install -y https://dl.google.com/linux/direct/google-chrome-stable_current_x86_64.rpm`

I needed to go this route, for whatever reason, `yum` didn't work.

**Reply by ZeeshanAhmadKhalil** [reply]
I have been looking for a solution for the last 5-6 hrs, thank you

**Reply by sarthakdevil** [reply]
> I'm getting this error:
> 
> ```
> Error: 
>  Problem: conflicting requests
>   - package google-chrome-stable-117.0.5938.88-1.x86_64 does not have a compatible architecture
>   - nothing provides ld-linux-x86-64.so.2()(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides ld-linux-x86-64.so.2(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides ld-linux-x86-64.so.2(GLIBC_2.3)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.10)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.11)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.14)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.15)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.16)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.3)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.3.2)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.3.3)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.3.4)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.4)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.6)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.7)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.8)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libc.so.6(GLIBC_2.9)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libdl.so.2(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libm.so.6(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libpthread.so.0(GLIBC_2.12)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libpthread.so.0(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libpthread.so.0(GLIBC_2.3.2)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libpthread.so.0(GLIBC_2.3.3)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides libpthread.so.0(GLIBC_2.3.4)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
>   - nothing provides librt.so.1(GLIBC_2.2.5)(64bit) needed by google-chrome-stable-117.0.5938.88-1.x86_64
> ```

same error anything to fix it

**Reply by ozbenh** [reply]
Are you trying to install on aarch64 maybe ? Google only provides chrome binaries for x86

**Comment by ozbenh** [comment]
Are you trying to install an x86_64 package on an aarch64 instance ?

**Comment by trivikr** [comment]
I attempted using `install` command in `@puppeteer/browsers` as mentioned in [Chrome for Testing](https://developer.chrome.com/blog/chrome-for-testing/) after installing node.

```console
$ node -v
v18.17.1

$ npx --yes @puppeteer/browsers install chrome@stable
Downloading chrome r117.0.5938.88 - 146.7 MB [====================] 100% 0.0s 
chrome@117.0.5938.88 /home/ec2-user/chrome/linux-117.0.5938.88/chrome-linux64/chrome
```

As of 2023-09-18, it fails with missing dependency `libatk-1.0.so.0`
```console
$ /home/ec2-user/chrome/linux-117.0.5938.88/chrome-linux64/chrome --version
/home/ec2-user/chrome/linux-117.0.5938.88/chrome-linux64/chrome: error while loading shared libraries: libatk-1.0.so.0: cannot open shared object file: No such file or directory
```

**Comment by ozbenh** [comment]
```
$ rpm -ql atk | grep \.so
/usr/lib64/libatk-1.0.so.0
/usr/lib64/libatk-1.0.so.0.23609.1
```
Try `sudo dnf install atk` ?

**Comment by JatB01** [comment]
Install missing dependencies:

```
sudo yum install libXcomposite libXdamage libXrandr libgbm libxkbcommon pango alsa-lib atk at-spi2-atk cups-libs libdrm -y
```

**Reply by Daniyal-thirtynorth** [reply]
Thanks Man worked me

**Reply by i00lii** [reply]
@JatB01, sorry for silly question - but how did you collect the list of 'missed' dependencies? [Here](https://github.com/umihico/docker-selenium-lambda/blob/main/Dockerfile) I found the similar list of depenedencies and it's quite similar to that you provide, but not the same.

**Reply by ChrisLane** [reply]
@i00lii
1. Attempt to run Chrome and get an error about a missing library
2. Figure out which package includes the library
3. Install the package
4. Repeat

**Reply by pjflopes** [reply]
I have used this in an Elastic beanstalk  running Node.js 20 running on 64bit Amazon Linux 2023/6.1.1.

there was a small change I've done, replace `libgbm` by `mesa-libgbm`

This is my configuration of packages in the _.ebextensions_ folder:

```
packages:
  yum:
    libXcomposite: []
    libXdamage: []
    libXrandr: []
    libxkbcommon: []
    pango: []
    alsa-lib: []
    atk: []
    at-spi2-atk: []
    cups-libs: []
    libdrm: []
    mesa-libgbm: []
```

**Reply by raphaelsetin** [reply]
@pjflopes Your `.ebextensions` configuration worked perfectly for me 🎉. Thank you!

---

### For others who stumble upon this:
I am currently using `"puppeteer": "^22.4.1"` and AMI `amazon/aws-elasticbeanstalk-amzn-2023.3.20240219.64bit-eb_nodejs20_amazon_linux_2023-hvm-2024-02-20T16-40`.

I also needed to install some fonts for my use case, which rendered the **additional** packages in my `.ebextensions` configuration file:
```
packages:
  yum:
    fontconfig: []
    dejavu-sans-fonts: []
    dejavu-sans-mono-fonts: []
    dejavu-serif-fonts: []
    liberation-mono-fonts: []
    liberation-sans-fonts: []
    liberation-serif-fonts: []
    google-noto-emoji-fonts: []
    google-noto-fonts-common: []
    google-noto-sans-fonts: []
    google-noto-serif-fonts: []
```

**Reply by limarodrigobr** [reply]
@pjflopes I LOVE YOU!!! GOD BLESS YOU!!!! THIS WORKS LIKE MAGIC!!!

**Reply by Vicopem01** [reply]
@pjflopes saved me hours of fixing, thank you

**Reply by cjosepha** [reply]
> I have used this in an Elastic beanstalk running Node.js 20 running on 64bit Amazon Linux 2023/6.1.1.
> 
> there was a small change I've done, replace `libgbm` by `mesa-libgbm`
> 
> This is my configuration of packages in the _.ebextensions_ folder:
> 
> ```
> packages:
>   yum:
>     libXcomposite: []
>     libXdamage: []
>     libXrandr: []
>     libxkbcommon: []
>     pango: []
>     alsa-lib: []
>     atk: []
>     at-spi2-atk: []
>     cups-libs: []
>     libdrm: []
>     mesa-libgbm: []
> ```

I have used this exact configuration for almost one year with success.
It appears this doesn't work anymore after I had to rebuild the ElasticBeanstalk environment, without changing the platform nor the app version (just changed the security key only). Now all deployments with the exact same app version fail and the server responds with 502 bad gateway.
There is nothing relevant in the eb-engine.log file that could help understand the issue.
Apparently something changed on the AWS side, do you experience the same issue?

**Comment by finalfantasia** [comment]
The following is what I did to have [Chrome for Testing](https://developer.chrome.com/blog/chrome-for-testing) working on an EC2 instance running Amazon Linux 2023. I hope this helps:

### With Node.js
```sh
# A hacky way to install all the package dependencies required for Chrome for Testing.
# See: https://github.com/GoogleChromeLabs/chrome-for-testing/issues/55
sudo dnf deplist https://dl.google.com/linux/direct/google-chrome-stable_current_x86_64.rpm \
  | grep provider \
  | sort --unique \
  | awk '{print $2}' \
  | xargs sudo dnf install --best --allowerasing --skip-broken --assumeyes --quiet >& /dev/null

npx --yes @puppeteer/browsers install chrome@stable \
  | awk '{print $2}' \
  | xargs -I {} sudo ln --symbolic {} /usr/local/bin/chrome

npx --yes @puppeteer/browsers install chromedriver@stable \
  | awk '{print $2}' \
  | xargs -I {} sudo ln --symbolic {} /usr/local/bin/chromedriver
```

### Without Node.js
```sh
# A hacky way to install all the package dependencies required for Chrome for Testing.
# See: https://github.com/GoogleChromeLabs/chrome-for-testing/issues/55
sudo dnf install --assumeyes --quiet findutils jq unzip
sudo dnf deplist https://dl.google.com/linux/direct/google-chrome-stable_current_x86_64.rpm \
  | grep provider \
  | sort --unique \
  | awk '{print $2}' \
  | xargs sudo dnf install --best --allowerasing --skip-broken --assumeyes --quiet >& /dev/null

CHROME_FOR_TESTING_RELEASE="$(curl --silent https://googlechromelabs.github.io/chrome-for-testing/last-known-good-versions-with-downloads.json | jq '.channels.Stable')"

CHROME_DOWNLOAD_URL="$(echo ""${CHROME_FOR_TESTING_RELEASE}"" | jq -r '.downloads.chrome[] | select(.platform == "linux64") | .url')"
curl --remote-name --silent "${CHROME_DOWNLOAD_URL}"
unzip -q chrome-linux64.zip
sudo ln --symbolic "${PWD}/chrome-linux64/chrome" /usr/local/bin/chrome

CHROMEDRIVER_DOWNLOAD_URL="$(echo ""${CHROME_FOR_TESTING_RELEASE}"" | jq -r '.downloads.chromedriver[] | select(.platform == "linux64") | .url')"
curl --remote-name --silent "${CHROMEDRIVER_DOWNLOAD_URL}"
unzip -q chromedriver-linux64.zip
sudo ln --symbolic "${PWD}/chromedriver-linux64/chromedriver" /usr/local/bin/chromedriver
```

**Reply by PavanMudigondaTR** [reply]
Could someone please confirm if this solution is working fine ?

**Reply by mizunoajigawakaru** [reply]
It worked, thank you.
(With Node.js)

**Reply by ION606** [reply]
this worked perfectly for me thankies ^~^
Also, if anyone's wondering how to use this in code, you can just do `await puppeteer.launch({product: "chrome", executablePath: "/usr/bin/chromium"})` (assuming you have nodejs installed). I should note that I was unable to put anything in `/usr/local/*`, which is why my path is just `/usr/bin/chromium`, but I'm not sure why....

**Comment by stewartsmith** [comment]
The flatpak of Chromium also works well. I've used this with remote X

**Comment by AMETemp** [comment]
> Could someone please confirm if this solution is working fine ?

extremely late but this worked for me

**Comment by tashrafy** [comment]
I've used bits and parts of the answers provided here as my use case was slightly different - I wanted to run `puppeteer` with chromium on an AL2023 Graviton instance (arm64).

Here is my `.ebextensions` config for installing chromium for puppeteer:

```
packages:
  yum:
    libXcomposite: []
    libXdamage: []
    libXrandr: []
    libxkbcommon: []
    pango: []
    alsa-lib: []
    atk: []
    at-spi2-atk: []
    cups-libs: []
    libdrm: []
    mesa-libgbm: []
    fontconfig: []
    dejavu-sans-fonts: []
    dejavu-sans-mono-fonts: []
    dejavu-serif-fonts: []
    liberation-mono-fonts: []
    liberation-sans-fonts: []
    liberation-serif-fonts: []
    google-noto-emoji-fonts: []
    google-noto-fonts-common: []
    google-noto-sans-fonts: []
    google-noto-serif-fonts: []

commands:
  chrome:
    command: |
        QTVER=5.15.9-9
        CHROMEVER=121.0.6167.85

        sudo yum install libgbm gtk3 libXcursor at-spi2-atk libXt xorg-x11-server-Xvfb xorg-x11-xauth dbus-glib dbus-glib-devel nss flac glibc-devel libX11 libXext libXi libXtst libXScrnSaver libva Xvfb soxr ipa-gothic-fonts liberation-fonts -y

        pushd /home/webapp
          rm -rf chrome-linux
          wget https://playwright.azureedge.net/builds/chromium/1099/chromium-linux-arm64.zip
          unzip chromium-linux-arm64.zip
          rm -f ./chromium-linux-arm64.zip
        popd

        sudo dnf -y install --best --allowerasing --skip-broken --assumeyes \
            https://kojipkgs.fedoraproject.org//packages/minizip/2.8.9/2.el8/aarch64/minizip-2.8.9-2.el8.aarch64.rpm \
            https://download-ib01.fedoraproject.org/pub/epel/9/Everything/aarch64/Packages/n/nss-mdns-0.15.1-3.1.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/gstreamer1-1.22.1-2.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/libcanberra-0.30-27.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/libcanberra-gtk3-0.30-27.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/sound-theme-freedesktop-0.8-17.el9.noarch.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/double-conversion-3.1.5-6.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/qt5-qtbase-$QTVER.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/qt5-qtbase-common-$QTVER.el9.noarch.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/qt5-qtbase-gui-$QTVER.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/glx-utils-8.4.0-12.20210504git0f9e7d9.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/pipewire-libs-0.3.67-2.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/fdk-aac-free-2.0.0-8.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/libldac-2.0.2.3-10.el9.aarch64.rpm \
            https://kojipkgs.fedoraproject.org//packages/chromium/$CHROMEVER/1.el9/aarch64/chromium-$CHROMEVER-1.el9.aarch64.rpm \
            https://kojipkgs.fedoraproject.org//packages/chromium/$CHROMEVER/1.el9/aarch64/chromium-common-$CHROMEVER-1.el9.aarch64.rpm \
            https://kojipkgs.fedoraproject.org//packages/chromium/$CHROMEVER/1.el9/aarch64/chromium-headless-$CHROMEVER-1.el9.aarch64.rpm \
            https://kojipkgs.fedoraproject.org//packages/chromium/$CHROMEVER/1.el9/aarch64/chromedriver-$CHROMEVER-1.el9.aarch64.rpm
        
        sudo chown -R webapp:webapp /home/webapp/chrome-linux/chrome
        export PUPPETEER_SKIP_CHROMIUM_DOWNLOAD true
        export CHROME_PATH=/home/webapp/chrome-linux/chrome
        export PUPPETEER_EXECUTABLE_PATH=/home/webapp/chrome-linux/chrome
```

The libraries installed are to ensure that `chromium` can run on the AL2023 Graviton instance.  You can also execute `ldd /path/to/chrome | grep not` to see what libraries are missing and required for `chromium`.

The install of chromium is coming from `playwright` - after hours of searching, I found [this thread that provides insight into obtaining the correct arm64 version of `chromium.`](https://github.com/puppeteer/puppeteer/issues/7740#issuecomment-1833202428)

You'll notice that I'm setting the permissions to use `webapp` as that's the same owner for the NodeJS application that will be running puppeteer and is deployed to the AL2023 instance.  

After doing all this, I was able to run `puppeteer` and have it use `chromium` on the AL2023 Graviton instance!

References: 
- https://github.com/aws/aws-graviton-getting-started/blob/main/software/ChromeAndPuppeteer.md
- https://medium.com/mockingbot/run-puppeteer-chrome-headless-on-ec2-amazon-linux-ami-6c9c6a17bee6

**Reply by neuronwave** [reply]
Perfect timing. I'd just given up on ARM and moved over to X86. Put this in and worked without a problem.

**Reply by cw-sarvesh** [reply]
Same here

**Reply by m3thom** [reply]
Thank you! You are my missing piece! I was able to install Chromium with `yarn` (npx should be the same):
`yarn playwright install chromium --with-deps`
After installation, the default installed path should look like this:
`/root/.cache/ms-playwright`
To use the Chromium binary, I have to specify it with something like this:
`/root/.cache/ms-playwright/chromium-1124/chrome-linux/chrome --version`

**Comment by sukhovv** [comment]
here is the full Beanstalk extension file that works for puppeteer on `Node.js 20 running on 64bit Amazon Linux 2023/6.1.6`, processor type: `x86_64`

file name: `.ebextensions/01_install_chrome.config`

```
packages:
  yum:
    libXScrnSaver: []
    atk: []
    gtk3: []
    cups-libs: []
    libXcomposite: []
    alsa-lib: []
    libXrandr: []
    pango: []
    libXcursor: []
    libXi: []
    libXtst: []
    libXinerama: []
    wget: []
    xorg-x11-server-Xvfb: []

commands:
  01_add_google_signing_key:
    command: |
      rpm --import https://dl-ssl.google.com/linux/linux_signing_key.pub

  02_setup_google_chrome_repo:
    command: |
      cat <<EOF > /etc/yum.repos.d/google-chrome.repo
      [google-chrome]
      name=google-chrome
      baseurl=http://dl.google.com/linux/chrome/rpm/stable/x86_64
      enabled=1
      gpgcheck=1
      gpgkey=https://dl-ssl.google.com/linux/linux_signing_key.pub
      EOF

  03_install_google_chrome:
    command: yum -y install google-chrome-stable

container_commands:
  01_set_chrome_headless:
    command: echo "alias chrome-headless='google-chrome --headless --no-sandbox --disable-gpu --remote-debugging-port=9222'" >> /etc/profile.d/chrome-headless.sh

```

**Reply by G0maa** [reply]
Hello @sukhovv, may I know your `Puppeteer` configuration?

I keep getting `Could not find Chrome` error, and it notes the cache path is `/home/webapp/.cache/puppeteer` (which is wrong) it's on `/root/.cache/puppeteer`.

EDIT:
I changed the default cache directory like [here](https://pptr.dev/guides/configuration#changing-the-default-cache-directory) and it now works... just need some missing fonts :grinning:

**Comment by neuronwave** [comment]
This is my config file for installing on Linux 2023 but on arm processor .ebextensions/10_arm_puppeteer_install.config

```
packages:
  yum:
    libXcomposite: []
    libXdamage: []
    libXrandr: []
    libxkbcommon: []
    pango: []
    alsa-lib: []
    atk: []
    at-spi2-atk: []
    cups-libs: []
    libdrm: []
    mesa-libgbm: []
    fontconfig: []
    dejavu-sans-fonts: []
    dejavu-sans-mono-fonts: []
    dejavu-serif-fonts: []
    liberation-mono-fonts: []
    liberation-sans-fonts: []
    liberation-serif-fonts: []
    google-noto-emoji-fonts: []
    google-noto-fonts-common: []
    google-noto-sans-fonts: []
    google-noto-serif-fonts: []

commands:
  chrome:
    command: |
        QTVER=5.15.9-9
        CHROMEVER=121.0.6167.85

        sudo yum install libgbm gtk3 libXcursor at-spi2-atk libXt xorg-x11-server-Xvfb xorg-x11-xauth dbus-glib dbus-glib-devel nss flac glibc-devel libX11 libXext libXi libXtst libXScrnSaver libva Xvfb soxr ipa-gothic-fonts liberation-fonts -y

        pushd /home/webapp
          rm -rf chrome-linux
          wget https://playwright.azureedge.net/builds/chromium/1099/chromium-linux-arm64.zip
          unzip chromium-linux-arm64.zip
          rm -f ./chromium-linux-arm64.zip
        popd

        sudo dnf -y install --best --allowerasing --skip-broken --assumeyes \
            https://kojipkgs.fedoraproject.org//packages/minizip/2.8.9/2.el8/aarch64/minizip-2.8.9-2.el8.aarch64.rpm \
            https://download-ib01.fedoraproject.org/pub/epel/9/Everything/aarch64/Packages/n/nss-mdns-0.15.1-3.1.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/gstreamer1-1.22.1-2.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/libcanberra-0.30-27.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/libcanberra-gtk3-0.30-27.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/sound-theme-freedesktop-0.8-17.el9.noarch.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/double-conversion-3.1.5-6.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/qt5-qtbase-$QTVER.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/qt5-qtbase-common-$QTVER.el9.noarch.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/qt5-qtbase-gui-$QTVER.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/glx-utils-8.4.0-12.20210504git0f9e7d9.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/pipewire-libs-0.3.67-2.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/fdk-aac-free-2.0.0-8.el9.aarch64.rpm \
            http://mirror.stream.centos.org/9-stream/AppStream/aarch64/os/Packages/libldac-2.0.2.3-10.el9.aarch64.rpm \
            https://kojipkgs.fedoraproject.org//packages/chromium/$CHROMEVER/1.el9/aarch64/chromium-$CHROMEVER-1.el9.aarch64.rpm \
            https://kojipkgs.fedoraproject.org//packages/chromium/$CHROMEVER/1.el9/aarch64/chromium-common-$CHROMEVER-1.el9.aarch64.rpm \
            https://kojipkgs.fedoraproject.org//packages/chromium/$CHROMEVER/1.el9/aarch64/chromium-headless-$CHROMEVER-1.el9.aarch64.rpm \
            https://kojipkgs.fedoraproject.org//packages/chromium/$CHROMEVER/1.el9/aarch64/chromedriver-$CHROMEVER-1.el9.aarch64.rpm

        sudo chown -R webapp:webapp /home/webapp/chrome-linux/chrome
        export PUPPETEER_SKIP_CHROMIUM_DOWNLOAD true
        export CHROME_PATH=/home/webapp/chrome-linux/chrome
        export PUPPETEER_EXECUTABLE_PATH=/home/webapp/chrome-linux/chrome
```


And this in my post deploy hook `./platform/hooks/postdeploy/install_puppeteer.sh`


```
#!/usr/bin/env bash
. /opt/elasticbeanstalk/deployment/env
cd /var/app
npm install puppeteer

```

and this is the js file I call to produce the pdfs `generate_pdf.js`

```
const puppeteer = require('puppeteer');

(async () => {
    const args = process.argv.slice(2);
    const url = args[0];
    const outputPath = args[1];

    const browser = await puppeteer.launch({
      args: ['--no-sandbox', '--disable-setuid-sandbox','--force-color-profile=srgb',
             '--font-render-hinting=none', '--disable-web-security'],
    });
    const page = await browser.newPage();
    await page.goto(url, { waitUntil: 'networkidle0' });

    // Modifing the page here
    await page.evaluate(_ => {
    // Capture all links that start with javascript on the href property
    // and change it to # instead.
    document.querySelectorAll('a')
      .forEach(a => {
        a.removeAttribute("href")
      })
    });

    await page.pdf({ path: outputPath,
                format: 'A4',
                displayHeaderFooter: true,
                printBackground: true,
                tagged: false,
                // preferCSSPageSize: true,
                margin: {
                    top: 30,
                    bottom: 40,
                    left: 40,
                    right: 40
                },
            });

    await browser.close();
})();
```


I then system call `node /path/to/script/generate_pdf.js 'https://example.com' path/to/output.pdf`

**Reply by ruurdkrekt** [reply]
Thank you so much, this was very helpful! 
Quick question, how do you make sure the installed version of chromium and all its dependencies stay up to date?

**Reply by neuronwave** [reply]
Excellent question and I don't know. I've chosen to ignore that critical issue for the moment ;)

**Reply by adlazare** [reply]
Recently asked AWS support this question - they suggest storing your own copies in S3 and installing from there when launching new instances (in case any of those links die, which they will based on my experience).

When you need to update your chromium + dependencies, it's just a process of finding links to all required libraries until you are able to complete the install. If someone knows a better way, please do tell.

**Comment by ankurk91** [comment]
Can some one suggest something about ARM64?
Could not make chromium running on arm :(

**Comment by johnthuss** [comment]
This works on al2023:
```
# Install
sudo rpm --import https://dl.google.com/linux/linux_signing_key.pub
sudo dnf install -y https://dl.google.com/linux/direct/google-chrome-stable_current_aarch64.rpm

# Usage example
google-chrome --headless --disable-gpu --dump-dom https://example.com

# Upgrade later
sudo dnf upgrade google-chrome-stable
```

Alternatively you can add the repo first and then install it by name:
```
sudo tee /etc/yum.repos.d/google-chrome.repo >/dev/null <<'EOF'
[google-chrome]
name=google-chrome
baseurl=https://dl.google.com/linux/chrome/rpm/stable/$basearch
enabled=1
gpgcheck=1
gpgkey=https://dl.google.com/linux/linux_signing_key.pub
EOF
sudo dnf install -y google-chrome-stable
```
